- Student 1 Name: Murshed SK
- Student 2 Name: Atyozzal Konwar

change the name of this notebook to  `name_1_name_2_notebook_??.ipynb` with *no spaces, no accents and no strange characters!* and where `??` stands for the number of the notebook you are working on.

# PPM Numerical Methods -- Numerical Methods for Physics

# Numerical methods: Root finding

# Root finding

## Bisection method

Use the bisection method to find the root of the function
$$ f(x) = \frac{1}{2} - e^{-x}$$
think carefully how to estimate the error to end the calculation.

In [ ]:
# Import necesary libraries for this notebook:
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import time  # Import the time module to check the time taken to complete the process

In [ ]:
# Define the function:
def f(x: float) -> float:
    func = 1/2 - np.exp(-x)
    return func

# Define the algorithm part of the Bisection method:
def bisection(a,b,tolerance):
    """
    Finds the root of a function using the bisection method.

    Args:
        a: The left endpoint of the interval.
        b: The right endpoint of the interval.
        tolerance: The maximum error allowed.

    Returns:
        True if a root is found within the tolerance, False otherwise.
    """
    if f(a)*f(b)>0:
        print("The root is not in between the interval.")
        return False
    else:
      iteration = 0 # Iteration is used to find the total number of iterations is needed to find the root.
      while (b-a)/2 >tolerance:
        c = (a+b)/2 # The mid point of the interval which is the new end point after iteration.
        if f(c) == 0:
          break # c is the root of the function.
        elif f(a)*f(c)<0:
          b = c # The root lies between 'a' and 'c'.
        else:
          a = c # The root lies between 'b' and 'c'.
        iteration += 1
    print(f"The root of the given function after {iteration}th itration is:{c:.4f}")
    return True


In [ ]:
# Main function to drive the program
def main():
    while True:
        try:
            a = float(input("Enter the lower bound of the interval (a): "))
            b = float(input("Enter the upper bound of the interval (b): "))
            tolerance = float(input("Enter the tolerance (maximum error): "))

            if a >= b:
                print("Error: The lower bound must be less than the upper bound. Please try again.")
                continue
            if tolerance <= 0:
                print("Error: Tolerance must be a positive number. Please try again.")
                continue

        except ValueError:
            print("Invalid input. Please enter correct numeric values.")
            print("Starting the process from the beginning. Enter the correct inputs this time.")
            print("")
            continue

        # Start the timer
        start_time = time.time()

        # Attempt to find the root:
        if bisection(a, b, tolerance):
            # Stop the timer
            end_time = time.time()
            print(f"Execution time: {end_time - start_time:.4f} seconds")
            break  # Exit the loop if a valid root is found.
        else:
            print("Please select another interval.")
            print("")

# Entry point of the program:
if __name__ == "__main__":
    main()


Enter the lower bound of the interval (a): q
Invalid input. Please enter correct numeric values.
Starting the process from the beginning. Enter the correct inputs this time.

Enter the lower bound of the interval (a): 1
Enter the upper bound of the interval (b): 2
Enter the tolerance (maximum error): 1e-6
The root is not in between the interval.
Please select another interval.

Enter the lower bound of the interval (a): 0
Enter the upper bound of the interval (b): 1
Enter the tolerance (maximum error): 1e-6
The root of the given function after 19th itration is:0.6931
Execution time: 0.0012 seconds


## False-position method


Use the false position method to find the root of the function
$$ f(x) = \frac{1}{2} - e^{-x}$$
and compare to the bisection method

In [ ]:
# Define the function:
def f(x: float) -> float:
    func = 1/2 - np.exp(-x)
    return func

# Define the algorithm part of the False Position method:
def falseposition(a,b,tolerance):
    """
    Performs the false position method to find the root of the function f.

    Args:
        a: The left endpoint of the interval.
        b: The right endpoint of the interval.
        tolerance: The desired tolerance (maximum error) for the root.

    Returns:
        True if a root is found within the tolerance, False otherwise.
    """
    if f(a)*f(b)>0:
        print("The root is not in between the interval.")
        return False
    else:
      iteration = 0 # iteration is used to find the total number of iterations is needed to find the root.
      while (b-a)/2 >tolerance:
        c = b - (f(b)*(b-a))/(f(b)-f(a)) # The new position of the interval which is the new end point after iteration.
        if f(c) == 0:
          break # c is the root of the function.
        elif f(a)*f(c)<0:
          b = c # The root lies between 'a' and 'c'.
        else:
          a = c # The root lies between 'b' and 'c'.
        iteration += 1
    print(f"The root of the given function after {iteration}th itration is:{c:.4f}")
    return True


In [ ]:
# Main function to drive the program
def main():
    while True:
        try:
            a = float(input("Enter the lower bound of the interval (a): "))
            b = float(input("Enter the upper bound of the interval (b): "))
            tolerance = float(input("Enter the tolerance (maximum error): "))

            if a >= b:
                print("Error: The lower bound must be less than the upper bound. Please try again.")
                continue
            if tolerance <= 0:
                print("Error: Tolerance must be a positive number. Please try again.")
                continue

        except ValueError:
            print("Invalid input. Please enter correct numeric values.")
            print("Starting the process from the beginning. Enter the correct inputs this time.")
            print("")
            continue

        # Start the timer
        start_time = time.time()

        # Attempt to find the root:
        if falseposition(a, b, tolerance):
            # Stop the timer
            end_time = time.time()
            print(f"Execution time: {end_time - start_time:.4f} seconds")
            break  # Exit the loop if a valid root is found.
        else:
            print("Please select another interval.")
            print("")

# Entry point of the program:
if __name__ == "__main__":
    main()


Enter the lower bound of the interval (a): q
Invalid input. Please enter correct numeric values.
Starting the process from the beginning. Enter the correct inputs this time.

Enter the lower bound of the interval (a): 1
Enter the upper bound of the interval (b): 2
Enter the tolerance (maximum error): 1e-6
The root is not in between the interval.
Please select another interval.

Enter the lower bound of the interval (a): 0
Enter the upper bound of the interval (b): 1
Enter the tolerance (maximum error): 1e-6
The root of the given function after 30th itration is:0.6931
Execution time: 0.0017 seconds


###*Comparison* between [`Bisection method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=lmuXFrC_z3Sr) and [`False Position method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=Orua1V4zz3S0&line=2&uniqifier=1).
We take same values for both case to compare these two method and run the code.
Those values are given below -

*   The lower bound of the interval (a) is : 0
*   The upper bound of the interval (b) is : 1
*   The tolerance (maximum error) is : 1e-6

After running each code, we observe the following -

|          | [`Bisection method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=lmuXFrC_z3Sr)     | [`False Position method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=Orua1V4zz3S0&line=2&uniqifier=1) |
|--------------|:-----------:|:------------:|
| Iterations | 19      | 30       |
| Execution time      | 0.0012 seconds  | 0.0017 seconds       |




From these results we are concluding that Bisection method is more effective (in the point of view of iterations and execution time) than the False Position method for the chosen parameters(values).







## The Newton-Raphson Method

Implement the Newton-Rapshon method to solve
$$ f(x) = \frac{1}{2} - e^{-x}$$
and compare to the bisection and false position methods

- Try different starting guess values, e.g. -1, 1, 5 and 30
- Comment

In [ ]:
# Define the function:
def f(x: float) -> float:
    return 1/2 - np.exp(-x)

# Define the derivative of the function:
def df(x: float) -> float:
    h = 1e-6  # The limit from the definition of Derivative.
    return (f(x + h) - f(x)) / h  # Definition of Derivative.

# Define the algorithm part of the Newton-Raphson method:
def newton_raphson(guess: float, tolerance: float, maximum_iterations) -> float:
    """
    Implements the Newton-Raphson method to find the root of a function.

    Args:
        guess: The initial guess for the root.
        tolerance: The desired accuracy of the solution.
        max_iterations: The maximum number of iterations.
    Returns:
        The approximate root of the function, or None if the method fails to converge.
    """
    iteration = 0  # Used to count the number of iterations.
    while iteration < maximum_iterations:  # Limit the number of iterations to prevent infinite loops.
        if abs(f(guess)) < tolerance:
            print(f"The root of the given function after {iteration} iterations is: {guess:.4f}")
            return guess  # Return the found root.

        derivative = df(guess)
        if derivative == 0:
            print("Derivative is zero. No solution found.")
            return None  # Avoid division by zero.

        new_guess = guess - f(guess) / derivative
        if abs(new_guess - guess) < tolerance:
            print(f"The root of the given function after {iteration} iterations is: {new_guess:.4f}")
            return new_guess  # Return the found root.

        guess = new_guess
        iteration += 1



In [ ]:
# Main function to drive the program
def main():
    while True:
        try:
            guess = float(input("Enter the initial guess: "))
            tolerance = float(input("Enter the tolerance (maximum error): "))
            maximum_iterations = int(input("Enter the maximum number of iterations: "))

            if tolerance <= 0:
                print("Error: Tolerance must be a positive number. Please try again.")
                continue
            if maximum_iterations <= 0:
                print("Error: Maximum number of iterations must be a positive integer. Please try again.")
                continue
        except ValueError:
            print("Invalid input. Please enter correct numeric values.")
            print("Starting the process from the beginning. Enter the correct inputs this time.")
            print("")
            continue

        # Start the timer
        start_time = time.time()

        # Attempt to find the root:
        if newton_raphson(guess, tolerance, maximum_iterations) is not None:
            # Stop the timer
            end_time = time.time()
            print(f"Execution time: {end_time - start_time:.4f} seconds")
            break  # Exit the loop if a valid root is found.
        else:
            print("Newton-Raphson method failed to converge.")
            print("Please make another guess.")
            print("")

# Entry point of the program:
if __name__ == "__main__":
    main()


Enter the initial guess: 1q
Invalid input. Please enter correct numeric values.
Starting the process from the beginning. Enter the correct inputs this time.

Enter the initial guess: 10
Enter the tolerance (maximum error): 1e-6
Enter the maximum number of iterations: 100


<ipython-input-6-5c3c3e4eaa04>:3: RuntimeWarning: overflow encountered in exp
  return 1/2 - np.exp(-x)
<ipython-input-6-5c3c3e4eaa04>:8: RuntimeWarning: invalid value encountered in scalar subtract
  return (f(x + h) - f(x)) / h  # Definition of Derivative.


Newton-Raphson method failed to converge.
Please make another guess.

Enter the initial guess: 10
Enter the tolerance (maximum error): 1e-6
Enter the maximum number of iterations: 100
Newton-Raphson method failed to converge.
Please make another guess.

Enter the initial guess: 1
Enter the tolerance (maximum error): 1e-6
Enter the maximum number of iterations: 100
The root of the given function after 3 iterations is: 0.6931
Execution time: 0.0013 seconds


###*Comparison* between [`Bisection method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=lmuXFrC_z3Sr), [`False Position method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=Orua1V4zz3S0&line=2&uniqifier=1) and [`Newton-Raphson method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=hbthxeJKz3S3).
We take same values for both case to compare these two method and run the code.
Those values are given below -


*   The tolerance (maximum error) is : 1e-6

After running each code, we observe the following -

|          | [`Bisection method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=lmuXFrC_z3Sr)     |[`False Position method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=Orua1V4zz3S0&line=2&uniqifier=1)| [`Newton-Raphson method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=hbthxeJKz3S3) |
|--------------|:-----------:|:------------------------:|:------------:|
| Iterations | 19      |30| 3       |
| Execution time      | 0.0012 seconds  |0.0017 seconds| 0.0013 seconds       |



From these results we are concluding that the execution time for the same tolerance in the Bisection method and the Newton-Raphson method is comparable, but not for the False Position method. Therefore, the Newton-Raphson method is the most effective and the False Position method is least effective method (in the point of view of iterations and execution time) for the chosen parameters(values).








## The Secant Method

Implement the Newton-Rapshon method to solve
$$ f(x) = \frac{1}{2} - e^{-x}$$
and compare to the bisection and false position methods

- Try different starting guess values, e.g. -1, 1, 5 and 30
- Comment

In [ ]:
# Define the function:
def f(x: float) -> float:
    return 1/2 - np.exp(-x)

def secant(x0, x1, tolerance, max_iterations):
    """
    Implements the secant method to find the root of a function.

    Args:
        x0: The first initial guess for the root.
        x1: The second initial guess for the root.
        tolerance: The desired accuracy of the solution.
        max_iterations: The maximum number of iterations.

    Returns:
        The approximate root of the function, or None if the method fails to converge.
    """
    iteration = 0
    while iteration < max_iterations:
        if abs(f(x1)) < tolerance:
            print(f"The root of the given function after {iteration} iterations is: {x1:.4f}")
            return x1

        if f(x1) - f(x0) == 0:
            print("Division by zero encountered. Method failed.")
            return None

        x2 = x1 - f(x1) * (x1 - x0) / (f(x1) - f(x0))

        if abs(x2 - x1) < tolerance:
            print(f"The root of the given function after {iteration} iterations is: {x2:.4f}")
            return x2

        x0 = x1
        x1 = x2
        iteration += 1

    print("Secant method failed to converge.")
    return None


In [ ]:
# Main function to drive the program
def main():
    while True:
        try:
            x0 = float(input("Enter the first initial guess: "))
            x1 = float(input("Enter the second initial guess: "))
            tolerance = float(input("Enter the tolerance (maximum error): "))
            max_iterations = int(input("Enter the maximum number of iterations: "))

            if tolerance <= 0:
                print("Error: Tolerance must be a positive number. Please try again.")
                continue
            if max_iterations <= 0:
                print("Error: Maximum number of iterations must be a positive integer. Please try again.")
                continue
        except ValueError:
            print("Invalid input. Please enter correct numeric values.")
            print("Starting the process from the beginning. Enter the correct inputs this time.")
            print("")
            continue

        # Start the timer
        start_time = time.time()

        # Attempt to find the root:
        if secant(x0, x1, tolerance, max_iterations) is not None:
            # Stop the timer
            end_time = time.time()
            print(f"Execution time: {end_time - start_time:.4f} seconds")
            break  # Exit the loop if a valid root is found.
        else:
            print("Secant method failed to converge.")
            print("Please make another guess.")
            print("")


if __name__ == "__main__":
    main()


Enter the first initial guess: 1
Enter the second initial guess: -1
Enter the tolerance (maximum error): 1e-6
Enter the maximum number of iterations: 100
The root of the given function after 6 iterations is: 0.6931
Execution time: 0.0014 seconds


###*Comparison* between [`Bisection method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=lmuXFrC_z3Sr), [`False Position method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=Orua1V4zz3S0&line=2&uniqifier=1) and [`Secant Method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=NAj4x7bkz3S4).
We take same values for both case to compare these two method and run the code.
Those values are given below -


*   The tolerance (maximum error) is : 1e-6

After running each code, we observe the following -

|          | [`Bisection method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=lmuXFrC_z3Sr)     |[`False Position method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=Orua1V4zz3S0&line=2&uniqifier=1)| [`Secant Method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=NAj4x7bkz3S4)|
|--------------|:-----------:|:------------------------:|:------------:|
| Iterations | 19      |30| 6      |
| Execution time      | 0.0012 seconds  |0.0017 seconds| 0.0014 seconds       |



From these results we are concluding that the execution time for the same tolerance in the Bisection method and the Secant method is comparable, but not for the False Position method. Therefore, the Secant method is the most effective and the False Position method is least effective method (in the point of view of iterations and execution time) for the chosen parameters(values).







## The Modified Secant Method

Implement the modified secant method and compare it to the other methods.

In [ ]:
#Define the function
def f(x: float) -> float:
    return 1/2 - np.exp(-x)

def modified_secant(x0, delta, tolerance, max_iterations):
    """
    Implements the modified secant method to find the root of a function.

    Args:
        x0: The initial guess for the root.
        delta: Perturbation factor.
        tolerance: The desired accuracy of the solution.
        max_iterations: The maximum number of iterations.

    Returns:
        The approximate root of the function, or None if the method fails to converge.
    """
    iteration = 0
    while iteration < max_iterations:
        if abs(f(x0)) < tolerance:
            print(f"The root of the given function after {iteration} iterations is: {x0:.4f}")
            return x0

        if f(x0 + delta * x0) - f(x0) == 0:
            print("Division by zero encountered. Method failed.")
            return None

        x1 = x0 - f(x0) * (delta * x0) / (f(x0 + delta * x0) - f(x0))

        if abs(x1 - x0) < tolerance:
            print(f"The root of the given function after {iteration} iterations is: {x1:.4f}")
            return x1

        x0 = x1
        iteration += 1


In [ ]:
# Main function to drive the program
def main():
    while True:
        try:
            x0 = float(input("Enter the initial guess: "))
            delta = float(input("Enter the perturbation factor (delta): "))
            tolerance = float(input("Enter the tolerance (maximum error): "))
            max_iterations = int(input("Enter the maximum number of iterations: "))

            if tolerance <= 0:
                print("Error: Tolerance must be a positive number. Please try again.")
                continue
            if max_iterations <= 0:
                print("Error: Maximum number of iterations must be a positive integer. Please try again.")
                continue
        except ValueError:
            print("Invalid input. Please enter correct numeric values.")
            print("Starting the process from the beginning. Enter the correct inputs this time.")
            print("")
            continue

        # Start the timer
        start_time = time.time()

        # Attempt to find the root:
        if modified_secant(x0, delta, tolerance, max_iterations) is not None:
            # Stop the timer
            end_time = time.time()
            print(f"Execution time: {end_time - start_time:.4f} seconds")
            break  # Exit the loop if a valid root is found.
        else:
            print("Modified secant method failed to converge.")
            print("Please make another guess.")
            print("")


if __name__ == "__main__":
    main()


Enter the initial guess: 1
Enter the perturbation factor (delta): 1e-6
Enter the tolerance (maximum error): 1e-6
Enter the maximum number of iterations: 100
The root of the given function after 3 iterations is: 0.6931
Execution time: 0.0003 seconds


###*Comparison* between [`Bisection method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=lmuXFrC_z3Sr), [`False Position method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=Orua1V4zz3S0&line=2&uniqifier=1) and [`Secant Method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=NAj4x7bkz3S4).
We take same values for both case to compare these two method and run the code.
Those values are given below -


*   The tolerance (maximum error) is : 1e-6

After running each code, we observe the following -

|          | [`Bisection method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=lmuXFrC_z3Sr)     |[`False Position method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=Orua1V4zz3S0&line=2&uniqifier=1)| [`Secant Method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=NAj4x7bkz3S4)|[`The Modified Secant  Method`](https://colab.research.google.com/drive/1fh12VNuD37e7NCDA6XHIo9Qo3F_M4nyM#scrollTo=lb-OX5oIz3S5)|
|--------------|:-----------:|:------------------------:|:------------:|:----:|
| Iterations | 19      |30| 6      |3|
| Execution time      | 0.0012 seconds  |0.0017 seconds| 0.0014 seconds |0.0003 seconds|



From these results we are concluding that the execution time for the same tolerance in the Bisection method and the Secant method is comparable, but not for the False Position method. Therefore, the Secant method is the most effective and the False Position method is least effective method (in the point of view of iterations and execution time) for the chosen parameters(values) among the first three methods.

`Overall` The Modified Secant method is the most effective and the False Position method is least effective method (in the point of view of iterations and execution time) for the chosen parameters(values).





---
###***Notes:***
We consider that instead of fixing the initial conditions (a, b, maximum error), it is better to give the inputs from the terminal. When giving the inputs we enter some absurd values (in our case 'q' instead of '1'). We did it because we think there might be input errors from the users (like '1' is just above 'q' in the QWERTY keyboard and sometimes there might be a typing error.) We also consider the range/interval where the root belong. So if the root is not in the chosen range, then we can again chose the range accordingly. In our case we chose the interval [1,2] and get that the root(viz. 0.6931) is not within the range. These parts are additional in the codes and probably not requied also for the asssignments. We did it because we wanted to extend our thinking of the possibility.


---